rust-symcrypt/src/ecurve.rs

```diff
@@ -0,0 +1,76 @@
+/* Friendly rust types for CurveTypes. */
+
+//use lazy_static::lazy_static;
+use crate::errors::SymCryptError;
+use symcrypt_sys;
+
+#[derive(Copy, Clone, PartialEq)]
+pub enum CurveType {
+    NistP256,
+    NistP384,
+    Curve25519,
+}
+
+pub(crate) fn convert_curve(curve: CurveType) -> symcrypt_sys::PCSYMCRYPT_ECURVE_PARAMS {
+    match curve {
+        CurveType::NistP256 => unsafe { symcrypt_sys::SymCryptEcurveParamsNistP256 }, // SAFETY: FFI calls
+        CurveType::NistP384 => unsafe { symcrypt_sys::SymCryptEcurveParamsNistP384 }, // SAFETY: FFI calls
+        CurveType::Curve25519 => unsafe { symcrypt_sys::SymCryptEcurveParamsCurve25519 }, // SAFETY: FFI calls
+    }
+}
+
+pub(crate) fn get_num_format(curve_type: CurveType) -> i32 {
```

Comment:
>get_num_format [](http://example.com/codeflow?start=14&length=14)

(Mentioned this on Teams, but copying it here for posterity)

I was trying to understand why you're using a different format for Curve25519. I looked it up and I guess Curve25519 typically uses LE. But the `SYMCRYPT_NUMBER_FORMAT` options are only used for serializing and deserializing, and do not affect SymCrypt's internal data structures. So either option should be fine for any curve.

I think this is fine as-is, but maybe leave a comment explaining why you wrote it this way.

Reply:
This is called for the following APIs:
SymCryptEckeySetValue, SymCryptEckeyGetValue and SymCryptEcDhSecretAgreement.

Each of which have the following input parameter:
SYMCRYPT_NUMBER_FORMAT  format,

Reply:
In the file: \onecore\ds\security\cryptoapi\ncrypt\msprim\ecc\ecc.c, I'm seeing the following for this parameter for the APIs listed above:

IsCurve25519(pKey->pParameters) ? SYMCRYPT_NUMBER_FORMAT_LSB_FIRST : SYMCRYPT_NUMBER_FORMAT_MSB_FIRST,

Reply:
We will be importing/exporting the 25519 public key using these APIs. What is the format on the wire, LSB or MSB?

Reply:
Ah, interesting. I don't know why BCrypt does that either. Possibly for the same reason I mentioned above, that Curve25519 conventionally uses LSB first. But I don't know if the format BCrypt/SymCrypt formats are directly compatible with those used by other implementations, e.g. OpenSSL. I'll ask Niels about this (or you can if you run into him first.)

Reply:
Confirmed that the code is written this way for TLS interop. The spec for 25519 defines it to use LE. Could you just add a comment indicating that?